ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/rtb/AppLovinRtbAdapter.java

```diff
@@ -0,0 +1,124 @@
+
+import android.app.Application;
+import android.text.TextUtils;
+import android.util.Log;
+
+import com.applovin.mediation.AppLovinUtils;
+import com.applovin.sdk.AppLovinSdk;
+import com.google.ads.mediation.sample.adapter.BuildConfig;
+import com.google.android.gms.ads.AdSize;
+import com.google.android.gms.ads.mediation.rtb.AdRenderingCallback;
+import com.google.android.gms.ads.mediation.rtb.BannerAd;
+import com.google.android.gms.ads.mediation.rtb.BannerEventListener;
+import com.google.android.gms.ads.mediation.rtb.InterstitialAd;
+import com.google.android.gms.ads.mediation.rtb.InterstitialEventListener;
+import com.google.android.gms.ads.mediation.rtb.RewardedAd;
+import com.google.android.gms.ads.mediation.rtb.RewardedEventListener;
+import com.google.android.gms.ads.mediation.rtb.RtbAdConfiguration;
+import com.google.android.gms.ads.mediation.rtb.RtbAdapter;
+import com.google.android.gms.ads.mediation.rtb.RtbConfiguration;
+import com.google.android.gms.ads.mediation.rtb.RtbSignalData;
+import com.google.android.gms.ads.mediation.rtb.SignalCallbacks;
+import com.google.android.gms.ads.mediation.rtb.VersionInfo;
+
+import java.util.List;
+
+/**
+ * Created by Thomas So on July 17 2018
+ */
+public class AppLovinRtbAdapter
+        extends RtbAdapter
+{
+    private static final String TAG = "AppLovinRtbAdapter";
+
+    // TODO: Why is this not in the base adapter?!
+    // @Override
+    // public void setUp()
+    // {
+    // AppLovinSdk.getInstance( new Application() ).initializeSdk();
+    // }
+
+    @Override
+    public void initialize()
+    {
+        AppLovinSdk.getInstance( new Application() ).initializeSdk();
+    }
+
+    @Override
+    public void updateConfiguration(final List<RtbConfiguration> list)
+    {
+        super.updateConfiguration( list );
+    }
+
+    @Override
+    public VersionInfo getSdkVersion()
+    {
+        String versionString = AppLovinSdk.VERSION;
+        String splits[] = versionString.split( "\\." );
+        int major = Integer.parseInt( splits[0] );
+        int minor = Integer.parseInt( splits[1] );
+        int patch = Integer.parseInt( splits[2] );
+
+        return new VersionInfo( major, minor, patch );
+    }
+
+    @Override
+    public VersionInfo getAdapterVersion()
+    {
+        String versionString = BuildConfig.VERSION_NAME;
+        String splits[] = versionString.split( "\\." );
+        int major = Integer.parseInt( splits[0] );
+        int minor = Integer.parseInt( splits[1] );
+        // Adapter versions have 2 patch versions. Multiply the first patch by 100.
+        int micro = Integer.parseInt( splits[2] ) * 100 + Integer.parseInt( splits[3] );
+
+        return new VersionInfo( major, minor, micro );
+    }
+
+    @Override
+    public void collectSignals(RtbSignalData rtbSignalData, SignalCallbacks signalCallbacks)
+    {
+        // TODO: I hope that we do not use the SDK Key and Context from here to initialize SDK / get bid token with...
```

Comment:
I don't think we would be initializing the SDK here. We initialize it in either `initialize()` or `setUp()`. 
According to AdMob's docs: 
> `collectSignals()` is called on a background thread, and has a timeout of 1 second.

So, we should not be doing any heavy duty work here and try to callback with a bid token ASAP.

Reply:
I understand, but we can't initialize our SDK in `initialize()` or `setUp()`(wherever that method is) b/c there is no `Context` parameter in those methods :/.